nixos/modules/services/networking/tinydns.nix

```diff
@@ -1,55 +1,279 @@
+        #!${pkgs.stdenv.shell} -e
+        cd ${cfg.rootPath}
+        (
+          flock -n 5 || exit 1
+          mv data data.backup
+          cat >data
+          if ${getBin pkgs.djbdns}/bin/tinydns-data ; then
+            rm data.backup
+          else
+            mv data.backup data
+          fi
+          # Do not allow zone transfers to run more often than once per second due to
+          # minimum DNS record time resolution.
+          sleep 1
+        ) 5>data-lock
+        '';
+      in optionalString (cfg.secondary.sshKey != "") ''
+        restrict,command="${tinydnsReceive}" ${cfg.secondary.sshKey}
+      '') config.services.tinydns;
+    };
+    users.groups.tinydns = {
+      gid = config.ids.gids.tinydns;
+    };
 
-    systemd.services.tinydns = {
+    systemd.services = (mapAttrs' (ip: cfg: nameValuePair "tinydns-${ip}" {
+      enable = cfg.enable;
       description = "djbdns tinydns server";
       wantedBy = [ "multi-user.target" ];
-      after = [ "network.target" ];
-      path = with pkgs; [ daemontools djbdns ];
+      after = [ "network-online.target" ];
+      requires = [ "network-online.target" ];
+      environment = {
+        IP = ip;
+        ROOT = cfg.rootPath;
+        UID = "${toString config.users.users.tinydns.uid}";
+        GID = "${toString config.users.groups.tinydns.gid}";
+        AXFR_JOBS = "2"; # Number of concurrent zone transfers to secondary servers.
+      };
+      path = with pkgs; [ coreutils djbdns ];
+      serviceConfig = {
+        # Drops privileges and chroots to ROOT by itself
+        ExecStart = "${getBin pkgs.djbdns}/bin/tinydns";
+        LimitDATA = mkDefault "40000000";
+        RestartSec = 1;
+        Restart = "always";
+      };
       preStart = ''
-        rm -rf /var/lib/tinydns
-        tinydns-conf tinydns tinydns /var/lib/tinydns ${config.services.tinydns.ip}
-        cd /var/lib/tinydns/root/
-        ln -sf ${pkgs.writeText "tinydns-data" config.services.tinydns.data} data
+        rm -rf $ROOT
+        mkdir -m750 -p $ROOT
+        chown ''${UID}:''${GID} $ROOT
+        cd $ROOT
+        ${if cfg.secondary.enable then "touch data" else "ln -sf ${pkgs.writeText "tinydns-data" cfg.data} data"}
         tinydns-data
       '';
-      script = ''
-        cd /var/lib/tinydns
-        exec ./run
+      reload = let
+        tinydnsSend = pkgs.writeText "tinydns-send.mk" (concatStrings (
+          (singleton ''
+            all: ${concatStringsSep " " (attrNames cfg.secondaries)}
+          '') ++ (mapAttrsToList (k: v: ''
+.PHONY: ${k}
+${k}::
+	-${getBin pkgs.openssh}/bin/ssh -a -x -S none -T -i ${v.sshKey} ${v.uri} <${cfg.rootPath}/data
+'') cfg.secondaries)));
+      in ''
+        cd $ROOT
+        ${optionalString (!cfg.secondary.enable) "ln -sf ${pkgs.writeText "tinydns-data" cfg.data} data"}
+        tinydns-data
+        ${optionalString (cfg.secondaries != {}) "${getBin pkgs.gnumake}/bin/make -j$AXFR_JOBS -f ${tinydnsSend}"}
```

Comment:
If you want to run these jobs in parallel, why not use `xargs -P` instead? Or is there another reason why you use make instead?